### PR TITLE
Everything implemented as requested.

### PR DESCRIPTION
**Pull request by shplss**

### CODE REVIEW

File: Highway.cpp [hunk 1]
```diff
@@ -13,7 +19,20 @@ void Highway::changeSpeed(int newSpeed)
...
+
+    if(auto* car = dynamic_cast<Car*>(v))
+    {
+        car->closeWindows();
+    }
+
+    if(auto* bike = dynamic_cast<Motorcycle*>(v))
+    {
+        bike->lanesplitAndRace(100);
+    }
```
Comment (matkatmusic) [inline Highway.cpp:35]: - You are only supposed to call the member function that doesn't evade the cops.  you are not supposed to set the speed.
- Also, if you pass in a `Car` instance, you're still checking if it's a `Motorcycle`.  use `if/else if`.
- Where is your check for the `SemiTruck` type?
- [ ] revise

File: Highway.cpp [hunk 2]
```diff
@@ -22,7 +41,20 @@ void Highway::addVehicleInternal(Vehicle* v)
...
+
+    if(auto* bike = dynamic_cast<Motorcycle*>(v))
+    {
+        bike->tryToEvade();
+    }
+
+    if(auto* truck = dynamic_cast<SemiTruck*>(v))
+    {
+        truck->pullOver();
+    }
```
Comment (matkatmusic) [inline Highway.cpp:57]: use `if/else if`
- [ ] revise

File: Highway.h [hunk 3]
```diff
@@ -3,14 +3,25 @@
 #include <vector>
 #include <algorithm>
 
-struct Highway
+#include "Vehicle.h"
+
+struct Highway : public Vehicle
```
Comment (matkatmusic) [inline Highway.h:8]: A `Highway` is not a `Vehicle`.
What made you think it was?
- [ ] revise

File: Highway.h [hunk 4]
```diff
@@ -3,14 +3,25 @@
...
 {
+    Highway();
+
+    ~Highway() override;
+    Highway(const Highway&) = default;
+    Highway& operator=(const Highway&) = default;
```
Comment (matkatmusic) [inline Highway.h:14]: You don't need any of these. remove them.
- [ ] revise

File: HighwayPatrol.cpp [hunk 5]
```diff
@@ -1,16 +1,20 @@
...
+HighwayPatrol::HighwayPatrol() : Vehicle("HighwayPatrol") { }
+
+HighwayPatrol::~HighwayPatrol() { }
 
 void HighwayPatrol::scanHighway(Highway* h)
 {
     std::cout << name << ": scanning highway for speeders" << std::endl;
 
-    for( size_t i = h->vehicles.size(); --i >= 0; )
+    for( int i = h->vehicles.size(); --i >= 0; )
```
Comment (matkatmusic) [inline HighwayPatrol.cpp:17]: This change from `size_t` to `int` is causing a warning. put it back.  and look up what `std::vector::size()` returns on https://cppreference.com/ before you make blind changes to code you didn't write.
- [ ] revise

File: HighwayPatrol.cpp [hunk 6]
```diff
@@ -27,8 +31,25 @@ void HighwayPatrol::pullOver( Vehicle* v, bool willArrest, Highway* h )
...
+
+        if(auto* bike = dynamic_cast<Motorcycle*>(v))
+        {
+            type = "MOTORCYCLE";
+        }
+
+        if(auto* truck = dynamic_cast<SemiTruck*>(v))
+        {
+            type = "TRUCK";
+        }
```
Comment (matkatmusic) [inline HighwayPatrol.cpp:49]: `if/else if`
- [ ] revise

File: HighwayPatrol.h [hunk 7]
```diff
@@ -1,11 +1,16 @@
 #pragma once 
 
 #include "Vehicle.h"
+#include "Highway.h"
```
Comment (matkatmusic) [inline HighwayPatrol.h:4]: You aren't using Highway in this file at all.  Forward declare.  include in the cpp
- [ ] revise

File: SemiTruck.h [hunk 8]
```diff
@@ -1,2 +1,14 @@
...
+struct SemiTruck : public Vehicle
+{
+    SemiTruck(const std::string& s);
+
+    ~SemiTruck() override;
+    SemiTruck(const SemiTruck&) = default;
+    SemiTruck& operator=(const SemiTruck&) = default;
+
+    void pullOver();
+};
```
Comment (matkatmusic) [inline SemiTruck.h:14]: You are missing a function for when the `SemiTruck` is added to the highway.
- [ ] revise

File: Highway.h (hunk 3 above)
Comment (shplss) [reply Highway.h:8]: Was a misread from one of the errors I got. Not that I thought it was a vehicle, but that it needed to be derived for other reasons.

File: HighwayPatrol.cpp (hunk 5 above)
Comment (shplss) [reply HighwayPatrol.cpp:17]: This actually was changed due to the original code not running properly due to an error, thus the change. It was not changed blindly.
As soon as this is changed back to it's original type, there's an error with exit status 139.

File: Highway.cpp [hunk 9]
```diff
@@ -13,7 +17,16 @@ void Highway::changeSpeed(int newSpeed)
...
+    if(auto* car = dynamic_cast<Car*>(v))
+    {
+        car->closeWindows();
+    } else if(auto* bike = dynamic_cast<Motorcycle*>(v))
+    {
+        bike->lanesplitAndRace(100);
+    } else if(auto* truck = dynamic_cast<SemiTruck*>(v))
+    {
+        truck->easeIn();
+    }
```
Comment (matkatmusic) [inline Highway.cpp:29]: curly braces go on their own line.
- [ ] revise

File: Highway.cpp [hunk 10]
```diff
@@ -22,7 +35,16 @@ void Highway::addVehicleInternal(Vehicle* v)
...
+    if(auto* car = dynamic_cast<Car*>(v))
+    {
+        car->tryToEvade();
+    } else if(auto* bike = dynamic_cast<Motorcycle*>(v))
+    {
+        bike->tryToEvade();
+    } else if(auto* truck = dynamic_cast<SemiTruck*>(v))
+    {
+        truck->pullOver();
+    }
```
Comment (matkatmusic) [inline Highway.cpp:47]: curly braces go on their own line.
- [ ] revise

File: HighwayPatrol.cpp [hunk 11]
```diff
@@ -27,8 +32,21 @@ void HighwayPatrol::pullOver( Vehicle* v, bool willArrest, Highway* h )
...
+        if(auto* car = dynamic_cast<Car*>(v))
+        {
+            type = "CAR";
+        } else if(auto* bike = dynamic_cast<Motorcycle*>(v))
+        {
+            type = "MOTORCYCLE";
+        } else if(auto* truck = dynamic_cast<SemiTruck*>(v))
+        {
+            type = "TRUCK";
+        }
```
Comment (matkatmusic) [inline HighwayPatrol.cpp:46]: curly braces go on their own line.
- [ ] revise

**Pull request merged**